datalad/distributed/drop.py

```diff
@@ -0,0 +1,804 @@
+            if p == dirpath or p in dirpath.parents:
+                return True
+    return False
+
+
+def _drop_dataset(ds, paths, what, reckless, recursive, recursion_limit, jobs):
+    lgr.debug('Start dropping for %s', ds)
+    # we know that any given path is part of `ds` and not any of its
+    # subdatasets!
+
+    # by-passing this completely with reckless=kill
+    if recursive and not reckless == 'kill':
+        # process subdatasets first with recursion
+        for sub in ds.subdatasets(
+                # must be resolved!
+                path=paths or None,
+                # nothing to drop with unavailable subdatasets
+                fulfilled=True,
+                # we can use the full recursion depth, only the first layer
+                # of calls to _drop_dataset() must/can have recursive=True
+                recursive=recursive,
+                recursion_limit=recursion_limit,
+                # start reporting with the leaves
+                bottomup=True,
+                result_xfm='datasets',
+                on_failure='ignore',
+                return_type='generator',
+                result_renderer='disabled'):
+            yield from _drop_dataset(
+                ds=sub,
+                # everything, the entire subdataset is matching a given path
+                paths=None,
+                what=what,
+                reckless=reckless,
+                recursive=False,
+                recursion_limit=None,
+                jobs=jobs)
+
+    if not ds.pathobj.exists():
+        # basic protection against something having wiped it out already.
+        # should not happen, but better than a crash, if it does
+        yield dict(
+            action='drop',
+            path=ds.path,
+            status='notneeded',
+            message="does not exist",
+            type='dataset',
+        )
+        return
+
+    if paths is not None and paths != [ds.pathobj] and what == 'all':
+        # so we have paths constraints that prevent dropping the full dataset
+        lgr.debug('Only dropping file content for given paths in %s, '
+                  'allthough instruction was to drop %s', ds, what)
+        what = 'filecontent'
+
+    repo = ds.repo
+    is_annex = isinstance(repo, AnnexRepo)
+
+    # first critical checks that might prevent further operation
+    had_fatality = False
+    for res in _fatal_pre_drop_checks(
+            ds, repo, paths, what, reckless, is_annex):
+        had_fatality = True
+        yield res
+    if had_fatality:
+        return
+
+    # next check must come AFTER the modification checks above, otherwise
+    # remove() could not rely on the modification detection above
+    if paths is not None and paths != [ds.pathobj] and what == 'datasets':
+        # so we have paths constraints that prevent dropping the full dataset
+        # there is nothing to do here, but to drop keys, which we must not
+        # done
+        return
```

Comment:
@mih

Something along these lines is what I mean. When do we end up here? AFAICS, when we have path(s) that are part of THIS dataset (as opposed to a subdataset) and we have `what='datasets'`. So, phrasing it like the comment ("constraining the drop of THIS dataset") is one way, but "trying to drop subdataset(s) where there are none" is another.

May be I just don't see how we can end up here in a way that this message would be misleading. I don't know.

```suggestion
        for p in paths:
            yield dict(action="drop",
                       path=p,
                       status="impossible",
                       message=f"what=datasets was given, but no dataset can be dropped from {p}",
                       type="file")
        return
```

Reply:
I am afraid this is not correct. It is perfectly possible that the paths you are reporting on have been originally used to drop datasets underneath it. `impossible` would be wrong, because something might have been done already, and `file` is a guess.

Reply:
> It is perfectly possible that the paths you are reporting on have been originally used to drop datasets underneath it.

Hm. I fail to produce that, but thanks for the hint. Trying to figure that out.